Automattic/wp-calypso
# Framework: Add project specific `boot`

The aim here is to strip boot.js down into something sane, so we can: 
 - Start understanding it better
 - Allow different projects, or even entry points, to use a different boot file
 - Isolate the WPCOM specific stuff
 - Strip out what we can
 - asyncRequire what we can

~~This pseudo-code mess was a result of @gziolo, @mtias and myself. It's meant as a plan/starting point. Do not review :)~~
It is now ready to review.

### New boot folder structure

```
boot/
	app.js
	common.js
	dev-modules.js
	project
		wordpress-com.js
```

* There is new entry point for Webpack client build: `client/boot/app.js`. We need to decide if we want to keep `index.js` or `app.js`. This change was introduced mostly to make files changed list nicer on GitHub.
* Everything needed to make Calypso framework work was moved to `common.js` file.
* All development tools are now located in `dev-modules.js` file and are behind environment check to make sure they are never bundled for production envs.
* All remaining code remains in project specific file `project/wordpress-com.js`. This open possibilities to create different project setups, i.e. jetpack, dekstop, woo, etc.

### Testing
1. Execute `make distclean` and `make run`.
2. Open [http://calypso.localhost:3000](http://calypso.localhost:3000) or https://calypso.live/?branch=add/project-specific-boot.
3. Make sure it still works.
4. Type `su` to make sure happiness support still works. You should see login dialog.
5. Go to `me`.
6. Update any field and save data.
7. Navigate to `/me/account` and make sure success notice disappears.
8. Change your display language and make sure page reloads properly.
9. Change it back your language of choice and make sure it works properly.
10. Open http://calypso.localhost:3000/theme/heart-and-style in incognito tab (logged out).
11. Make sure it loads properly and SSR works as intended.
12. Open http://calypso.localhost:3000/me in incognito tab (logged out).
13. Make sure it renders only masterbar.

__Also we need to make sure that issue fixed in #12605 by @aduth won't show up again.__

Verify that the app boots while logged in, in Safari Private Mode and your preferred browser.

1. Log in at [WordPress.com](https://wordpress.com)
2. Navigate to [http://calypso.localhost:3000](http://calypso.localhost:3000)
3. Note that the Reader loads without issue

Verify new content of [README](https://github.com/Automattic/wp-calypso/blob/dd22b70e35a4842ddbcdc58e022d7a0cdf3b3330/client/boot/README.md) file in the `client/boot` folder.

### TODO
* [x] Update README file in `client/boot` folder. 
* [x] Fix SSR when logged out (no errors):
![screen shot 2017-03-28 at 15 21 21](https://cloud.githubusercontent.com/assets/699132/24407102/40bac80c-13ca-11e7-933c-f90d133fdb65.png)
* [x] Load project specific boot file dynamically (another PR?).
* [ ] Rename back `client/boot/app.js` to `client/boot/index.js` after PR gets accepted? (to confirm)
